design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+	// If this field is not set, an Order is not in progress.
+	// This field may point to a failed or inactive Order.
+	// It is not sufficient to check for the presence of this field in order to
+	// determine whether an order is in progress.
+	// +optional
+	OrderRef *LocalObjectReference `json:"orderRef,omitempty"`
+}
+```
+
+### Controllers
+
+Two new controllers will be built to reconcile the new `Order` and `Challenge`
+types.
+
+This two controllers will take the majority of the logic that used to exist in
+the `Prepare` function of the ACME issuer.
+
+#### Order controller
+
+When an order is created, all of the spec fields must be defined.
+The Order controller will be responsible for handling the entirety of an ACME
+order.
+The 'spec' describes an order to be created. The controller will keep the 'status'
+fields up to date by periodically syncing with the ACME server.
+
+* The Order controller will attempt to create a new order with the ACME server
+if the `status.url` field is not set. It will copy details of the order back
+onto the Order resoure, to save subsequent calls to the ACME server when not required.
+
+* If the orders 'state' is 'final' (i.e. one of valid, invalid or expired) then
+the controller will take no further action and return. (TODO: should we delete challenge
+resources for this order here too?)
+
+* For each authorization on the Order, the order controller will select a challenge
+type to use to solve the authorization (based on the Issuer config and Order's
+SolverConfig) and then ensure a Challenge resource for each authorization exists.
+
+* The order controller will check the status of the Challenge resources that are
+related to it, and if any of them have changed state, it will 'resync' the order
+status with the ACME server.
+
+* If the order is 'ready', the order controller will attempt to 'finalize' the
+order using the CSR as specified on the Order resource (i.e. `spec.csr`).
+After this finalization has completed, the order status will be updated and the
+order's `status.certificateURL` field will be set.
+
+#### Challenge controller
+
+The challenge controller is responsible for:
+
+* presenting challenges using a configured solver (i.e. a dns01 provider, or by
+creating resources in order to solve http01 challenges).
+
+* performing the 'self-check' to ensure the challenge record has propegated
+
+* 'accepting' the authorization once the self check is passing.
+
+* keeping the `status` field up to date with details of the challenge so that
+the Order controller can make decisions based on the state of challenges.
+
+One area to highlight, is the behaviour of the Challenge controller wrt. challenges
+vs authorizations.
+
+Whilst this controller works with a single ACME *Challenge* only, in order to
+avoid introducing a third resource type (i.e. `Authorization`), the Challenge
+controller is responsible for accepting the **authorization** associated with
```

Comment:
What does accepting an authorization mean in this context? Choosing the authorization based on the challenge-type?

Reply:
The Order controller creates Challenge resources which also contain the 'AuthorizationURL' field (i.e. we boil every ACME Authorization resource down to a single Challenge resource, ignoring the other possible challenge types - we pick which challenge/type to use based on the user configuration).

This note was to clarify which controller (Order or Challenge) is responsible for actually accepting the *authorization* using that challenge - which as the sentence says, is the Challenge controllers responsibility.

The Challenge controller will then update the Challenge resource with the final state of the *Authorization* after calling AcceptAuthorization.